### PR TITLE
feat: new command href and now ssp gen uses caching to pull catalog from remote

### LINKED ISSUE #557: Provide a way for href's to redirect to models within the trestle workspace
The href's in OSCAL files should correspond to where the files will ultimately be placed - but at the same time it would be useful for trestle to follow href's if files within the workspace are being worked on.

A simple approach would be to create a lookup table mapping href values to model types/names in the trestle workspace.  Perhaps a command like:

trestle href add catalog '../../../nist_catalog.json' nist_catalog

Then if trestle attempts to follow an href link it would first check if it is in the lut.

This may need to be linked to assemble if a certain directory structure is expected in the assembled models.

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

Fixed caching so that it handles windows drive letters better and handles timeout to pull from remote automatically.
Changed href resolution to allow either an absolute uri, either local or remote, that can be resolved, or something of form trestle://catalogs/mycat/catalog.json within the project.
Added new command href to change the import used by a catalog accordingly.
href is a general command rather than author command - but perhaps should be author subcommand if it is only relevant to ssp generation.

closes #557

### CODE REVIEW

File: trestle/utils/fs.py [hunk 1]
```diff
@@ -480,3 +486,21 @@ def text_files_equal(path_a: pathlib.Path, path_b: pathlib.Path) -> bool:
...
     return True
+
+
+def strip_drive_letter(file_path: str) -> Tuple[str, str]:
+    r"""If string starts with e.g. D:\\foo just return /foo along with the drive letter."""
+    drive_string_match = re.match(const.WINDOWS_DRIVE_URI_REGEX, file_path)
+    if drive_string_match is not None:
+        drive_string = drive_string_match[0]
+        drive_letter = drive_string[0]
+        return file_path.replace(drive_string, '/'), drive_letter
```
Comment (butler54) [inline trestle/utils/fs.py:497]: I would embed this as a class method into the cachine code as it's the only place it should be required.

File: trestle/utils/fs.py [hunk 2]
```diff
@@ -480,3 +486,21 @@ def text_files_equal(path_a: pathlib.Path, path_b: pathlib.Path) -> bool:
...
+    return file_path, ''
+
+
+def time_since_modification(file_path: pathlib.Path) -> datetime.timedelta:
+    """Get time since last modification."""
+    if not file_path.exists():
+        raise TrestleError(f'time since modification requested for non existent file {file_path}')
+    last_modification = datetime.datetime.fromtimestamp(file_path.stat().st_mtime)
```
Comment (butler54) [inline trestle/utils/fs.py:505]: this method as well. Keep it collected with the caching code.

**Review by butler54 — CHANGES_REQUESTED** [review]
I've left a number of comments through the code. 
There are a couple of ones I think we need to do:
1) Trestle import should be refactored to use the remote cache. This would be a nice change?
2) The caching will need some quite extensive documentation (which was why I thought about refactoring import at the same time)
3) I would like to do some  integration testing, however, that is a separate issue.

File: trestle/utils/fs.py [hunk 3]
```diff
@@ -450,17 +454,19 @@ def allowed_task_name(name: str) -> bool:
...
+    model_type: Type[OscalBaseModel],
+    expiration_seconds: int = const.DAY_SECONDS
+) -> common_types.TopLevelOscalModel:
+    """Load model from href string via cache if remote."""
+    # If the href is not of form trestle:// then return cached version
+    if re.match(const.TRESTLE_HREF_REGEX, href):
+        model_path = trestle_root / href[len(const.TRESTLE_HREF_HEADING):]
+        return model_type.oscal_read(model_path)
+    fetcher = cache.FetcherFactory.get_fetcher(trestle_root, href, expiration_seconds)
+    return fetcher.get_oscal(model_type)
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/utils/fs.py]: Done

File: trestle/core/commands/href.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+from trestle.core.commands.command_docs import CommandPlusDocs
+from trestle.utils import fs
+
+logger = logging.getLogger(__name__)
+
+
+class HrefCmd(CommandPlusDocs):
+    """Change href of import in profile to point to catalog in trestle project."""
+
+    name = 'href'
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/commands/href.py:33]: Done

File: trestle/core/commands/href.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    def _init_arguments(self) -> None:
+        logger.debug('Init arguments')
+        self.add_argument(
+            '-n', '--name', help='Name of trestle profile to modify (just its name).', type=str, required=True
+        )
+
+        self.add_argument(
+            '-hr', '--href', help='New href of form trestle://catalogs/mycat/catalog.json.', type=str, required=True
+        )
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/commands/href.py:43]: Added comments to the change_import_refs indicating future extensions.

File: trestle/core/commands/href.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    def _run(self, args: argparse.Namespace) -> int:
+        logger.debug('Entering trestle href.')
+
+        log.set_log_level_from_args(args)
+
+        if 'name' in args and args.name:
+            profile_name: str = args.name
+        else:
+            logger.warning('No profile name specified for command href.')
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/commands/href.py]: Done

File: trestle/core/commands/href.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            The original href is not checked and will be overwritten.
+
+        Future work:
+            Allow multiple imports with matching hrefs.
+            Allow href to point to profile in trestle.
+        """
+        trestle_root = fs.get_trestle_project_root(effective_cwd)
+        if trestle_root is None:
+            logger.warning(f'Effective directory is not in a trestle project: {effective_cwd}')
+            return 1
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/commands/href.py]: Done

File: trestle/core/remote/cache.py [hunk 8]
```diff
@@ -231,7 +255,7 @@ def __init__(self, trestle_root: pathlib.Path, uri: str, refresh: bool = False,
         if self._username and (self._password == '' or self._password is None):
             logger.error(
                 'Malformed URI, username found but valid password not found '
-                f'via environment variable in URL {self._uri}'
+                f'via environment variable in Uri {self._uri}'
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/remote/cache.py]: Done

File: trestle/core/remote/cache.py [hunk 9]
```diff
@@ -278,28 +302,27 @@ def _sync_cache(self) -> None:
...
-                logger.debug(f'HTTPSFetcher._sync_cache() request.json() method failed for url {self._url}: {err}')
-                logger.error(f'HTTPSFetcher sync failed, JSON error from getting url {self._url}: {err}')
+                logger.debug(f'HTTPSFetcher._do_fetch() request.json() method failed for url {self._url}: {err}')
+                logger.error(f'HTTPSFetcher fetch failed, JSON error from getting url {self._url}: {err}')
                 raise TrestleError(f'Cache update failure with expected JSON via HTTPS: {self._url} ({err})')
             else:
-                self._inst_cache_path.write_text(json.dumps(result))
+                self._cached_object_path.write_text(json.dumps(result))
         else:
             raise TrestleError(f'GET returned code {response.status_code}: {self._uri}')
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/remote/cache.py]: Now loading as text and saving as text.  Let me know if that needs to be changed.  No json errors being reported.

File: trestle/utils/fs.py (hunk 1 above)
Comment (fsuits) [reply trestle/utils/fs.py]: Done

File: trestle/utils/fs.py (hunk 2 above)
Comment (fsuits) [reply trestle/utils/fs.py]: Done

**Review by butler54 — APPROVED** [review]
LGTM - see a couple of small comments, however, squash merge when ready.

File: tests/trestle/conftest.py [hunk 10]
```diff
@@ -101,7 +101,7 @@ def sample_catalog():
 
 @pytest.fixture(scope='function')
 def sample_profile():
```
[the remark this answers is not in the document]
Comment (fsuits) [reply tests/trestle/conftest.py:103]: ok

File: tests/trestle/core/commands/href_test.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    # confirm new href is correct
+    new_profile: profile.Profile = profile.Profile.oscal_read(profile_path)
+    assert new_profile.imports[0].href == new_href
+
+    # restore orig href to confirm models are otherwise equivalent
+    new_profile.imports[0].href = orig_href
+    assert test_utils.models_are_equivalent(new_profile, sample_profile)
+
+
+def test_describe_failures(tmp_path: pathlib.Path, keep_cwd: pathlib.Path, sample_profile: profile.Profile) -> None:
```
[the remark this answers is not in the document]
Comment (fsuits) [reply tests/trestle/core/commands/href_test.py]: Done

File: tests/trestle/core/remote/cache_test.py [hunk 12]
```diff
@@ -34,17 +35,22 @@
 from trestle.utils import fs
 
 
+def as_file_uri(path: str) -> str:
+    """Convert sample non-existent path to file:/// and add drive letter if windows."""
+    # Correct usage would start with / and leaving off should cause errors with cache
+    prefix = '/C:' if platform.system() == const.WINDOWS_PLATFORM_STR else ''
+    return 'file://' + prefix + path
+
+
```
[the remark this answers is not in the document]
Comment (fsuits) [reply tests/trestle/core/remote/cache_test.py]: Fixed

File: tests/trestle/core/remote/cache_test.py [hunk 13]
```diff
@@ -338,35 +282,94 @@ def test_fetcher_bad_uri(tmp_trestle_dir: pathlib.Path, monkeypatch):
...
+    fetcher._expiration_seconds = 10
+    # should fetch because doesn't have it yet
+    assert fetcher._update_cache()
+    assert fetcher._cached_object_path.exists()
+
+    # should not fetch since it is too soon
+    assert not fetcher._update_cache()
+
+    # wait a bit
+    time.sleep(12)
```
[the remark this answers is not in the document]
Comment (fsuits) [reply tests/trestle/core/remote/cache_test.py]: Now 5s.  Don't want it to fail if there is a delay somewhere.

File: trestle/core/commands/href.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        trestle can load it.
+
+        If the catalog is already at the link referred to by the href as a valid URI or absolute file path then no
+        change is needed.  But if the catalog is being worked on in the same trestle directory as the profile,
+        the href should be modified to something like trestle://catalogs/my_catalog/catalog.json
+
+        This change only needs to be made once to the profile while the profile is being used to generate SSP's
+        from the local catalog, but if the final profile is released the href would need to be changed to the
+        intended final location of the catalog.
+
```
[the remark this answers is not in the document]
Comment (fsuits) [reply trestle/core/commands/href.py]: Done

**Review by butler54 — APPROVED** [review]
LGTM now

**Pull request merged**